protocols/request-response/src/handler.rs

```diff
@@ -140,24 +224,26 @@ where
                 // An event is reported to permit user code to react to the fact that
                 // the remote peer does not support the requested protocol(s).
                 self.pending_events
-                    .push_back(Event::OutboundUnsupportedProtocols(info));
-            }
-            StreamUpgradeError::Apply(e) => {
-                log::debug!("outbound stream {info} failed: {e}");
+                    .push_back(Event::OutboundUnsupportedProtocols(message.request_id));
             }
+            StreamUpgradeError::Apply(e) => void::unreachable(e),
             StreamUpgradeError::Io(e) => {
-                log::debug!("outbound stream {info} failed: {e}");
+                log::debug!(
+                    "outbound stream for request {} failed: {e}, retrying",
+                    message.request_id
+                );
+                self.requested_outbound.push_back(message);
```

Comment:
Not sure I still agree with this behaviour actually. I think @oblique wanted to have this reported as an error. That is probably a better idea than naively retrying!